d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -257,21 +257,25 @@ private void readyHandler()
       _log.warn("Unexpected state, ready called on null or backed off ADS stream!");
       return;
     }
-    // confirm ready state to neglect spurious callbacks; we'll get another callback whenever it is ready again.
-    if (_adsStream.isReady())
+    // Confirm ready state to neglect spurious callbacks; we'll get another callback whenever it is ready again.
+    // Also confirm ready timeout future is not null to avoid notifying multiple times.
+    if (!_adsStream.isReady() || _readyTimeoutFuture == null)
     {
-      // if the ready timeout future is non-null, a reconnect notification hasn't been sent yet.
-      if (_readyTimeoutFuture != null)
-      {
-        // timeout task will be cancelled only if it hasn't already executed.
-        boolean cancelledTimeout = _readyTimeoutFuture.cancel(false);
-        _log.info("ADS stream ready, cancelled timeout task: {}", cancelledTimeout);
-        _readyTimeoutFuture = null; // set it to null to avoid repeat notifications to subscribers.
-        _xdsClientJmx.incrementReconnectionCount();
-        notifyStreamReconnect();
-      }
-      _xdsClientJmx.setIsConnected(true);
+      return;
+    }
+
+    // timeout task will be cancelled only if it hasn't already executed.
+    boolean cancelledTimeout = _readyTimeoutFuture.cancel(false);
+    _log.info("ADS stream ready, cancelled timeout task: {}", cancelledTimeout);
+    _readyTimeoutFuture = null; // set it to null to avoid repeat notifications to subscribers.
+    if (_retryRpcStreamFuture != null)
+    {
+      _log.info("ADS stream reconnected after {} milliseconds",
+          _retryRpcStreamFuture.getDelay(TimeUnit.MILLISECONDS));
+      _retryRpcStreamFuture = null;
+      _xdsClientJmx.incrementReconnectionCount();
```

Comment:
All these conditions seem to be precluded by the `isInBackoff` check above at line 255? Also, `getDelay` seems to return the still remaining delay and not the elapsed time.

Reply:
> precluded by the isInBackoff check above at line 255
not really. This is not for checking backoff but just to clean _retryRpcStreamFuture for the case when the retry future is executed (and the reconnect stream is ready), but the future itself is not cleaned (previously reconnectionCount metric was incremented for initial connection as well, which was not a reconnect).

> getDelay seems to return the still remaining delay
whoops, that's right. Removed, thanks.